WebService/v2/Models/DeviceModelApiModel/DeviceModelApiModel.cs

```diff
@@ -0,0 +1,198 @@
+            if (this.CloudToDeviceMethods != null && this.CloudToDeviceMethods.Count > 0)
+            {
+                result.CloudToDeviceMethods = new Dictionary<string, Script>();
+                foreach (KeyValuePair<string, DeviceModelSimulationScript> method in this.CloudToDeviceMethods)
+                {
+                    var fieldValue = method.Value.ToServiceModel();
+                    result.CloudToDeviceMethods.Add(method.Key, fieldValue);
+                }
+            }
+
+            return result;
+        }
+
+        // Map service model to API model
+        public static DeviceModelApiModel FromServiceModel(DeviceModel value)
+        {
+            if (value == null) return null;
+
+            var result = new DeviceModelApiModel
+            {
+                ETag = value.ETag,
+                Id = value.Id,
+                Version = value.Version,
+                Name = value.Name,
+                Description = value.Description,
+                created = value.Created,
+                modified =value.Modified,
+                Protocol = value.Protocol.ToString(),
+                Simulation = DeviceModelSimulation.FromServiceModel(value.Simulation)
+            };
+
+            foreach (var property in value.Properties)
+            {
+                result.Properties.Add(property.Key, property.Value);
+            }
+
+            foreach (var message in value.Telemetry)
+            {
+                result.Telemetry.Add(DeviceModelTelemetry.FromServiceModel(message));
+            }
+
+            if (value.CloudToDeviceMethods?.Count > 0)
+            {
+                foreach (var method in value.CloudToDeviceMethods)
+                {
+                    result.CloudToDeviceMethods.Add(method.Key, DeviceModelSimulationScript.FromServiceModel(method.Value));
+                }
+            }
+
+            return result;
+        }
+
+        public void ValidateInputRequest(ILogger log)
+        {
+            const string NO_ETAG = "The custom device model doesn't contain an ETag";
+            const string NO_PROTOCOL = "The device model doesn't contain a protocol";
+            const string NO_ID = "The device model doesn't contain an id";
+            const string ZERO_TELEMETRY = "The device model has zero telemetry";
+
+            // A custom device model must contain a ETag
+            if (this.Type == "CustomModel" && this.ETag == String.Empty)
+            {
+                log.Error(NO_ETAG, () => new { deviceModel = this });
+                throw new BadRequestException(NO_ETAG);
+            }
+
+            // A device model must contain a protocol
+            if (this.Protocol == String.Empty)
+            {
+                log.Error(NO_PROTOCOL, () => new { deviceModel = this });
+                throw new BadRequestException(NO_PROTOCOL);
+            }
+
+            // A device model must contain an Id
+            if (this.Id == String.Empty)
```

Comment:
remove this, it's ok if the user doesn't send it, because we use the ID in the URL anyway

Reply:
Done.